lib/pgp.ml

```diff
@@ -0,0 +1,602 @@
+        else
+          get_old_length cs header_code
+      in
+      match length_infos with
+      | Error _ -> Error Fatal
+      | Ok (header_length, packet_length) -> (
+        match of_int tag with
+        | Ok packet_type ->
+          Ok (header_length, {packet_type; packet_length; is_new})
+        | Error _ -> Error (Header (header_length + packet_length)) )
+  end
+
+  module Id = struct
+    type t =
+      { name : string
+      ; email : string }
+    [@@deriving ord, eq, show]
+
+    let decode cs =
+      let id = Cstruct.to_string cs in
+      let sep_id = String.split_on_char '<' id in
+      let name = String.concat "<" (List.rev (List.tl (List.rev sep_id))) in
+      let email = List.nth sep_id (List.length sep_id - 1) in
+      {name; email = String.sub email 0 (String.length email - 1)}
+  end
+
+  module Public_key = struct
+    module Public_key_value = struct
+      type t =
+        | Rsa of Rsa.Public.t
+        | Dsa of Dsa.Public.t
+        | Elgamal of Elgamal.Public.t
+      [@@deriving ord, eq, show]
+    end
+
+    type t =
+      { version : int
+      ; creation_time : int32
+      ; validity_period : int option
+      ; algo : Algo.Public.t
+      ; public_key : Public_key_value.t }
+    [@@deriving ord, eq, show]
+
+    let decode_public_key (algo : Algo.Public.t) packet ~version =
+      let offset =
+        (*A public key packet has another header*)
+        match version with
+        | 3 ->
+          Ok 8
+          (*and a version 3 public key packet also contains a validity period*)
+        | 4 -> Ok 6
+        | _ -> Error "Public key packet has a bad version"
+      in
+      offset >>= fun off ->
+      match algo with
+      | Rsa_enc_sign
+      | Rsa_enc_only
+      | Rsa_sign_only ->
+        let (cs, key) = Rsa.Public.decode packet ~off in
+        Ok (cs, Public_key_value.Rsa key)
+      | Dsa ->
+        let (cs, key) = Dsa.Public.decode packet ~off in
+        Ok (cs, Dsa key)
+      | Elgamal_enc_only ->
+        let (cs, key) = Elgamal.Public.decode packet ~off in
+        Ok (cs, Elgamal key)
+      | Ec
+      | Ecdsa
+      | Unknown ->
+        Error ("Unsupported algorithm: " ^ Algo.Public.name algo)
+
+    let decode packet =
+      let version = Cstruct.get_uint8 packet 0 in
+      let creation_time = Cstruct.BE.get_uint32 packet 1 in
+      let packet_infos =
+        match version with
+        | 4 -> Ok (packet, None)
+        | 2
+        | 3 ->
+          let time = Cstruct.BE.get_uint16 packet 5 in
+          let cs = Cstruct.shift packet 2 in
+          Ok (cs, Some time)
+        | _ -> Error "Bad version of public key packet."
+      in
+      let algo = Algo.Public.of_int (Cstruct.get_uint8 packet 5) in
+      match packet_infos with
+      | Ok (public_packet, validity_period) ->
+        decode_public_key algo public_packet ~version >|= fun (cs, key) ->
+        (cs, {version; creation_time; validity_period; algo; public_key = key})
+      | Error _ -> Error "error"
```

Comment:
What about propagating the error message instead of overriding it?  This would make debugging easier, no?